### PR TITLE
Allow users to provide their own edge IDS to PropertyGraph

### LINKED ISSUE #2565: [FEA] Support adding `edge_ids` to PG in add_edge_data
**Describe the solution you'd like and any additional context**

Currently in PG [add_edge_data](https://github.com/rapidsai/cugraph/blob/cc057589799366b44eacf8b6f4a335223e4927c2/python/cugraph/cugraph/structure/property_graph.py#L478-L505)  we assign `edge_ids` on our own and the user cant add its own if they want to. 

**Context**

This feature is useful especially when converting b/w a graph from external  framework like DGL to cuGraph.  I think it might be worth-while to support adding `edge_ids`  directly in the below API so that the users can cleanly b/w frameworks. 

  

**Proposed API**
```python
   def add_edge_data(self,
                      dataframe,
                      vertex_col_names,
                      edge_id_col_name=None,
                      type_name=None,
                      property_columns=None
                      ):
        """ 
 ```

### PR DESCRIPTION
**Pull request by eriknw**
Closes #2565

Should we try to avoid creating conflicting edge IDs (for example, when user doesn't provides _all_ edge IDs).  Should we expose `last_edge_id`.

### CODE REVIEW

**Review by VibhuJawa — CHANGES_REQUESTED** [review]
The overall changes look good  but I think we should ask the user provide  edge_ids if they do it once and  also expose last_edge_id.

https://github.com/rapidsai/cugraph/pull/2757#issue-1390279530

**Comment by eriknw** [comment]
Case 1:
- User adds data with edge ids
- then user adds data w/o edge ids

Case 2:
- User adds data w/o edge ids
- then user adds data with edge ids

@VibhuJawa, do I understand your comment correctly that you want to support case 2, but not case 1?

We can handle case 1 by taking the max of the user-provided edge ids (as is currently done in this PR).  Not supporting this lets us avoid taking the max.

We could support case 2 by trusting the user.  Giving them the last (i.e., max) edge id could be a way to give them the tool to do this.  Exposing this value once again requires taking the max of the user-provided edge ids.

So, basically, all options are available to us:
- don't support case 1 or 2
  - no need to call max
- support case 1 only (current state of PR)
  - calls max, probably no need to expose last_edge_id
- support case 2 only
  - only call max if we also expose last edge id
- support cases 1 and 2
  - calls max, can expose last_edge_id

**Comment by VibhuJawa** [comment]
> support case 1 only (current state of PR)
calls max, probably no need to expose last_edge_id

I think this good for now .  The usecase i think this is useful is when we want to create new edges like adding self-loops etc . 

Can we create a feature request for other cases so that we can track and prioritize them as need arises.

**Review by VibhuJawa — APPROVED** [review]
LGTM

**Comment by rlratzel** [comment]
We had further discussion about case 1/case 2 offline, captured [here](https://github.com/rapidsai/cugraph/issues/2565#issuecomment-1265940964).

**Review by rlratzel — COMMENTED** [review]
Looks good, I just had one question, then I'll re-review after the changes from [this conversation](https://github.com/rapidsai/cugraph/issues/2565#issuecomment-1265940964) are done.

File: python/cugraph/cugraph/dask/structure/mg_property_graph.py [hunk 1]
```diff
@@ -571,17 +582,30 @@ def add_edge_data(self,
...
+            )
+            tmp_df[self.edge_id_col_name] = 1
+            tmp_df[self.edge_id_col_name] = (
+                tmp_df[self.edge_id_col_name].cumsum() + starting_eid
+            )
+            tmp_df = tmp_df.set_index(self.edge_id_col_name)
+            tmp_df = tmp_df.persist()
+            self.__last_edge_id = starting_eid + len(tmp_df.index)
+        else:
+            tmp_df.persist()
```
[the remark this answers is not in the document]
Comment (eriknw) [reply python/cugraph/cugraph/dask/structure/mg_property_graph.py:598]: ha, I just caught (and fixed) this too (and I thought I had already fixed this, but I think merging with squashed PRs isn't as smart as merging non-squished PRs, so maybe it snuck back in).  `tmp_df = tmp_df.persist()` is the correct spelling.  `tmp_df.persist()` without assignment is wasted work.

It's recommended to persist after setting index, and I think sometimes before setting index too.  I think we'll want to look more closely at usage of `persist` and operations like `set_index` in the coming months as we scale to larger data sets.

**Review by rlratzel — APPROVED** [review]
LGTM, especially the added test coverage.

**Comment by rlratzel** [comment]
@gpucibot merge

File: python/cugraph/cugraph/dask/structure/mg_property_graph.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (eriknw) [reply python/cugraph/cugraph/dask/structure/mg_property_graph.py]: See: https://docs.dask.org/en/stable/dataframe-best-practices.html#persist-intelligently
